chia/consensus/blockchain.py

```diff
@@ -1081,49 +1081,79 @@ async def get_block_generator(
-            # (as long as we're in the main chain)
-            result = await self.block_store.get_generators_at(block.transactions_generator_ref_list)
-        else:
-            # First tries to find the blocks in additional_blocks
-            curr = block
-            additional_height_dict = {}
-            while curr.prev_header_hash in additional_blocks:
-                prev: FullBlock = additional_blocks[curr.prev_header_hash]
-                additional_height_dict[prev.height] = prev
-                curr = prev
-
-            peak: Optional[BlockRecord] = self.get_peak()
-            prev_block_record = await self.get_block_record_from_db(curr.prev_header_hash)
-            reorg_chain: Dict[uint32, bytes32] = {}
-            if prev_block_record is not None and peak is not None:
+        generators: Dict[uint32, bytes] = {}
+
+        # The block heights in the transactions_generator_ref_list don't
+        # necessarily refer to the main chain. The generators may be found in 3
+        # different places. The additional blocks, a fork of the chain (but in
+        # the database) or in the main chain.
+
+        #              * <- block
+        # additional : |
+        #              * <- peak of fork (i.e. we have not
+        #              | :  validated blocks past this height)
+        #              | :
+        # peak -> *    | : reorg_chain
+        #          \   / :
+        #           \ /  :
+        #            *  <- fork point
+        #         :  |
+        #  main   :  |
+        #  chain  :  |
+        #         :  |
+        #         :  * <- genesis
+
+        generator_refs = set(block.transactions_generator_ref_list)
+
+        # traverse the additional blocks (if any) and resolve heights into
+        # generators
+        curr = block
+        to_remove = []
+        while curr.prev_header_hash in additional_blocks:
+            prev: FullBlock = additional_blocks[curr.prev_header_hash]
+            if prev.height in generator_refs:
+                if prev.transactions_generator is None:
+                    raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
+                generators[prev.height] = bytes(prev.transactions_generator)
+                to_remove.append(prev.height)
+            curr = prev
+        for i in to_remove:
+            generator_refs.remove(i)
+
+        if len(generator_refs) > 0:
+            # If the block immediately before the additional blocks is not part of
+            # the main chain, we're on a fork, and we need to find the fork point
+            prev_b = self.try_block_record(curr.prev_header_hash)
+            if prev_b is None:
+                prev_b = await self.get_block_record_from_db(curr.prev_header_hash)
+            assert prev_b is not None
+            if self.height_to_hash(prev_b.height) != curr.prev_header_hash:
+                peak: Optional[BlockRecord] = self.get_peak()
+                assert peak is not None
```

Comment:
why do you need to use the peak ?
if the block is not in height_to_hash that means the fork is before that block, so traversing from the peak back to that block seems redundant 
you would need to pull it out of DB of its not cached anyway and if the its way in the past using the peak will make us do db calls for more blocks